### ISSUE TITLE
Implement compiled method field watch support via read barrier and write barrier (JIT part)

### ISSUE THREAD

**Issue by cathyzhyi** [eclipse/openj9#2668]
As things stand currently under full speed debug, when an agent requests a field to be watched we cannot JIT compile the method because we cannot respect the hooks for reporting accesses of the respective field. This limitation will be lifted by using the rdbar / rdbari/ wrbar/ wrbari opcodes to represent reads/writes which have requested fields of certain types to be watched.

The JIT will generate read/write barriers for such fields of the correct type and the optimizer will treat them as normal loads/writes. At codegen time we will insert a C helper call to notify the VM via a hook that a field will be read/written and reduce the read/write barrier to the load/store.

In summary the following parts needs to be done in JIT:                  

- [x] Adding rdbar/rdbari opcodes: As outlined in [1] and discussed in [2] (video available on YouTube) I am adding a new IL opcode rdbar / rdbari to represent read barriers. Implemented in https://github.com/eclipse/omr/pull/2875 and https://github.com/eclipse/openj9/pull/2649 

- [x] Adding wrtbar/wrtbari opcodes: As outlined in [3] and implemented in and [4] [5]

- [x] Codegen supports on P, Z and X: see https://github.com/eclipse/openj9/issues/2668#issuecomment-414756016  for details. See [6] for X codegen work

- [x] general changes for functional correctness of the new opcodes
- [x] changes in optimizer for performance of the new opcodes
       

[1] https://github.com/eclipse/omr/issues/2812
[2] https://github.com/eclipse/omr/issues/2817
[3] https://github.com/eclipse/omr/issues/2948
[4] https://github.com/eclipse/omr/pull/3027 
[5] https://github.com/eclipse/openj9/pull/3138
[6] https://github.com/eclipse/openj9/pull/3134

@andrewcraik @gita-omr @fjeremic @gacholio

**Comment by cathyzhyi** [comment]
For each read and write barrier node, JIT needs to test J9ClassHasWatchedFields bit for the J9Class and if set call the appropriate vmhelper to report the event. All event reporting should take place before the read or write. During the event handler for a read, the value being read may be changed, so you should not cache the value across the report call. 
VM provides JIT codegen with 4 helpers to report the instance/static field read/writes:  jitReportInstanceFieldRead, jitReportInstanceFieldWrite, jitReportStaticFieldRead, jitReportStaticFieldWrite. Data blocks for wrapping some of the arguments are needed because the vmhelpers can only accept at most 3 arguments from JIT. The data block contains arguments known at compile time (field offset, calling method etc.) and can be stored into data snippets. Helpers arguments and data block definitions are the following: 

**jitReportInstanceFieldRead**
arg1 pointer to static data block
arg2 object being read

**jitReportInstanceFieldWrite**
arg1 pointer to static data block
arg2 object being written to
arg3 pointer to value being written

**jitReportStaticFieldRead**
arg1 pointer to static data block

**jitReportStaticFieldWrite**
arg1 pointer to static data block
arg2 pointer to value being written
The two data block formats are:
 
typedef struct J9JITWatchedInstanceFieldData {
    J9Method *method;        /* Currently executing method */
    UDATA location;            /* Bytecode PC index */
    UDATA offset;            /* Field offset (not including header) */
} J9JITWatchedInstanceFieldData;
 
typedef struct J9JITWatchedStaticFieldData {       
    J9Method *method;        /* Currently executing method */
    UDATA location;            /* Bytecode PC index */
    void *fieldAddress;        /* Address of static field storage */
    J9Class *fieldClass;    /* Declaring class of static field */
} J9JITWatchedStaticFieldData;                     

For unresolved case, to avoid patching fields in data snippets on different codegen we are proposing to always call a special helper what will do both the resolving and the field report. This will introduce performance penalty for unresolved fields read/write but it should be rare. @gacholio GAC will help with this part and we will finalize the design if it's feasible.

In the case of an object field being written, the helpers do not assume that the JIT is keeping the object value alive.

**Comment by cathyzhyi** [comment]
> For unresolved case, to avoid patching fields in data snippets on different platforms we are proposing to always call the resolving helper first and store the result into the data snippet without patching assembly code. 

It turns out in the current world, we can't call the resolving helper directly from jitted code without using picBuilder because the data resolving helpers expect different stack shape than normal helpers. Now there are 2 ways to solve this. One is adding a new picBuilder which doesn't do any patching but just return the result. The other approach is VM providing JIT with directly callable version of the resolving helpers which expect a normal stack shape. 

I think the second approach is cleaner and less likely to go wrong but would like to make sure all platforms agree on the same approach before we go ahead. @fjeremic @gita-omr What do you guys think about this?

**Comment by gacholio** [comment]
If (as I expect), we go forward with this then I see no reason that the static field resolves need to tag the result (they low-tag the result if the field class is currently being initialized) - save one instruction in each unresolved. I'm not sure the picbuilder ones do anything with the tag either (other than remove it).

**Comment by fjeremic** [comment]
Just spoke to @cathyzhyi about this.

> I think the second approach is cleaner and less likely to go wrong but would like to make sure all platforms agree on the same approach before we go ahead. @fjeremic @gita-omr What do you guys think about this?

I totally agree. Having a new VM helper with helper call linkage (preserves all registers) would be easiest for everyone IMO. In pseudo-code the resolved and unresolved cases look like this in pseudo-code:

```
Resolved:
---------------------------------------------------------------------
Call field report VM helper passing field report data snippet in the literal pool as argument
Load field at offset X
---------------------------------------------------------------------



Unresolved:
---------------------------------------------------------------------
if (offset in field report struct is == 0 branch) {
    Call new VM field resolve helper returning offset X
    Store X into field report data snippet in the literal pool
}

Call field report VM helper passing field report data snippet in the literal pool as argument
Call VM resolve PicBuilder helper Label1
Label2 (mainline):
Load field at offset X

...
...

Label1 (in PicBuilder):
Call VM resolve helper to determine the value of X
Patch X into mainline load
NOP the call to VM resolve PicBuilder helper Label1
Branch back to Label2 (mainline)
---------------------------------------------------------------------
```

The only thing we're missing is `Call field report VM helper passing field report data snippet in the literal pool as argument` from the VM side.

**Comment by cathyzhyi** [comment]
> The only thing we're missing is Call field report VM helper passing field report data snippet in the literal pool as argument from the VM side

we have those field report VM helpers already. did you mean `Call new VM field resolve helper returning offset X` ?

**Comment by gacholio** [comment]
OK, based on 2 votes, I'll go ahead and implement the new helpers.

**Comment by gacholio** [comment]
I see a problem with this solution - the `<clinit>` case is being ignored, and since you're not calling the resolve every time (you are effectively patching, it's just data rather than code), the static field will be accessible to other threads too early.

From what I've seen, this may already be an error in the normal PicBuilder, at least on X. I'm constructing a test case to see if this is true.

**Comment by gacholio** [comment]
OK, looks like I'm wrong about the `<clinit>` case being incorrect in general. Not completely sure how to handle this in this case

Also occurs that two fields need to be patched for static - the address and the declaring class (which is not necessarily the referenced class in the constant pool).

In case I'm being vague - the `<clinit>` case is this:

Thread 1

Refer to class X
`<clinit>` for X
during `<clinit>`, compile and run method m()
in m(), read a static on X

In this case, the read is allowed to proceed despite the class not being initialized (because it is being initialized on this thread). The caveat is that the JIT must not patch away the resolution check in the compiled code in case Thread 2 comes along and executes compiled m() - Thread 2 has to call the resolve helper in order to wait for Thread 1 to complete the `<clinit>` before it can read the static from X.

So somehow, we need to be able to keep calling the direct resolve helper in this case. Easiest thing to do would be to add a new field to the static data snippet to indicate resolution, rather than using address == 0.

**Comment by gacholio** [comment]
Summing up the rant, I think we need these things:

- extra field in static data snippet for resolution
- direct resolve helper does return the `<clinit>` tag in the address
- return both address and declaring class for static fields (how does this work in the normal unresolved case?)

**Comment by DanHeidinga** [comment]
Do you need the extra data or can you use the Class's initialized status to block in the reporting helper if the current thread is different than the initializing thread?

**Comment by gacholio** [comment]
That sounds good to me - it's a quick check for the expected case (or cases).

I am still wondering where the JIT gets the declaring class from in the unresolved `<clinit>` case - the VM resolution will not patch the RAM CP with the information.

**Comment by gacholio** [comment]
There's 4 (static/instance, setter/getter).

Is `jitResolveClassFromStaticField` always called after successfully resolving the field? It performs a full resolve, regardless of `<clinit>` status. Calling this from a thread which is initializing the declaring class would resolve the interpreter CP entry, opening us up to the multithread issue.

As far as I know, writes of objects to static fields require a barrier in gencon, so I'm not sure how you could get away without having it in hand in the unresolved case.

**Comment by gacholio** [comment]
I will add a direct-call helper for jitResolveClassFromStaticField as well.

**Comment by gacholio** [comment]
Turns out my feeling that something was amiss here was justified:

```
public class Holder {
	public static String i;

	static {
		System.out.println("<clinit> on " + Thread.currentThread() + " sleeping 5000 between writes");
		Clinit.write("better never see this");
                try { Thread.sleep(5000); } catch(Throwable t) {}
		Clinit.write("this is the real value");
		System.out.println("<clinit> complete");
	}
}

public class Clinit {
	public static void write(String a) {Holder.i = a;}
	public static String read() { return Holder.i; }

	public static void main(String args[]) throws Throwable {
		Thread t1 = new Thread() {
			public void run() {
				System.out.println("t1 running sleeping 1000");
				try { Thread.sleep(1000); } catch(Throwable t) {}
				System.out.println("t1 got value: " + read());
			}
		};
                Thread t2 = new Thread() {
                        public void run() {
				System.out.println("t2 running sleeping 2000");
                                try { Thread.sleep(2000); } catch(Throwable t) {}
                                System.out.println("t2  got value: " + read());
                        }
                };
		t1.start();
		t2.start();
		System.out.println("main thread got value: " + read());
	}
}
```

```
/team/gac/80/jre/bin/java "-Xjit:count=0,limit={Clinit.write*}" Clinit
t1 running sleeping 1000
t2 running sleeping 2000
<clinit> on Thread[main,5,main] sleeping 5000 between writes
t1 got value: better never see this
t2  got value: better never see this
<clinit> complete
main thread got value: this is the real value
```
```
/team/gac/80/jre/bin/java -Xint Clinit
t1 running sleeping 1000
t2 running sleeping 2000
<clinit> on Thread[main,5,main] sleeping 5000 between writes
<clinit> complete
main thread got value: this is the real value
t2  got value: this is the real value
t1 got value: this is the real value
```

**Comment by gacholio** [comment]
`jitResolveClassFromStaticField` is resolving the RAM CP entry, which is allowing the other threads in early. I believe this is a simple enough fix - I'll look into it.

According to @cathyzhyi, we don't need a direct version of this helper.

**Comment by gacholio** [comment]
If we want this test in, I'd want to firm it up a bit, in particular, synchronizing threads with a monitor rather than sleeps. Also, only one forked thread is required.

**Comment by gacholio** [comment]
New helpers: https://github.com/eclipse/openj9/pull/2795

**Comment by cathyzhyi** [comment]
@gacholio For the new helpers we still need to pass in the return address right?

**Comment by gacholio** [comment]
As they're written right now, yes, but I could remove the 3rd parameter and peek the calling address (like most other runtime helpers do).

**Comment by cathyzhyi** [comment]
@gacholio removing the third argument would be great if there aren't too much work. Passing return address from jit code is a little bit awkward.

**Comment by gacholio** [comment]
#2795 updated.

**Comment by gacholio** [comment]
#2795 is now merged.

**Comment by gacholio** [comment]
@gita-omr has asked for clarification on the static field/method resolve result tagging.

The tag on the static resolve return values is used to prevent bypassing the class initialization check that is required before allowing access to static data.

If the tag is set, this indicates that the current thread is initializing the class of the static data. The field/method are allowed to be resolved, but must not prevent other threads from waiting for the class initialization to complete.

In the interpreter, this is accomplished by not updating the constant pool entry after the resolution, forcing all other threads into the resolve code where the thread will block waiting for class inititlization to complete.

In the JIT, what it should mean is that the code cache is not patched with the resolved value (unless you can think of something more clever). If the code is patched, other threads will be able to access the static data without the class init check.

At the moment, the JIT appears to simply discard the tag and continue.

**Comment by gacholio** [comment]
Field test above added in #3800

**Comment by gita-omr** [comment]
Please feel free to add links to my comment above.

**Comment by gita-omr** [comment]
@pshipton somehow this is not part of any milestone... We are targeting Power part for 0.14. I will open(find) a Power sub-issue for this.

**Comment by cathyzhyi** [comment]
The support for x and optimizer is done for this item. 
Z is most done and has 2 WIP PR under review https://github.com/eclipse/openj9/pull/4743, https://github.com/eclipse/omr/pull/3571. 
P hasn't really started much yet https://github.com/eclipse/openj9/issues/3422

**Comment by gita-omr** [comment]
P is targeting 0.14 so @AlenBadel is actively working on it. Note that it's not a simple port from Intel. We had to implement new type of snippets and there are other Power specifics.

**Comment by dchopra001** [comment]
#4743 and eclipse/omr#3571 introduce several changes in the fieldwatch area with respect to the JIT code generators. One of the main tasks was to make the implementation common across all the code generators and only implement platform specific routines where needed. I have outlined key points about the codegen design below:

### Where are the evaluators defined
Most of the read/write barrier evaluators (ex. `rdbarEvaluator` or `wrtbarEvaluator`) are now implemented in the common codegen since they have identical implementations. Specifically, they all invoke a fieldwatch helper to generate instructions that will report a fieldwatch event at runtime, and then invoke the appropriate `loadEvaluator` or `storeEvaluator` to generate the actual load or store instruction. 

The evaluators that are **not** in the common codegen are:
- `irdbar`, `irdbari`, `ardbar` and `ardbari` evaluators: 
    - Since Software Concurrent Scavenge also uses these IL opcodes (on at least one platform), they are implemented uniquely on each platform. These evaluators will also invoke the common fieldwatch helper before generating CS specific code.

- `dwrtbar*` and `fwrtbar*` evaluators: 
    - These have to be implemented separately because their corresponding store evaluators are mapped to uniquely named functions on some platforms (i.e. their implementations are unique for each platform).

### Implementation details for the evaluators
The `rdbar` and `wrtbar` IL designates one of its children as a "side-effect" child (The remaining children, if any, are used by the `loadEvaluator` or `storeEvaluator`). The "side-effect" child provides the information needed to report the fieldwatch event. Thus the side-effect child is only useful when fieldwatch is enabled. 
In general, each of the `rdbar` evaluators are implemented like below:

```
rdbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
{
TR::Register *sideEffectReg = cg->evaluate(node->getFirstChild() /* sideEffectChild */);
// Call helper routine to handle fieldwatch case
rdWrtbarHelperForFieldWatch(node, cg, sideEffectReg, NULL);
cg->decReferenceCount(node->getFirstChild() /* sideEffectChild */);
TR::TreeEvaluator::loadEvaluator(node, cg);
```
All `rdbar` and `wrtbar` evaluators evaluate the "side-effect" child and then provide its result (in a `TR::Register`) to a common fieldwatch helper routine (`rdWrtbarHelperForFieldWatch`). 

In the case of `wrtbar` nodes, we also evaluate the child that holds the new value that is being written. This is because that information is needed when we report the fieldwatch event. Similar to the "side-effect" child, the value child is evaluated and its result is provided in a register to the common fieldwatch helper. The `wrtbarEvaluators` generally look like this:

```
wrtbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
{
TR::Register *sideEffectReg = cg->evaluate(node->getFirstChild() /* sideEffectChild */);
TR::Register *valueReg = cg->evaluate(node->getSecondChild() /* valueChild */);
// Call helper routine to handle fieldwatch case
rdWrtbarHelperForFieldWatch(node, cg, sideEffectReg, valueReg);
cg->decReferenceCount(node->getFirstChild() /* sideEffectChild */);
// skip decrementing the second child and let the store evaluator do it.
TR::TreeEvaluator::storeEvaluator(node, cg);
```

**Note on reference counts**: Generally speaking if a routine evaluates a node, then it must also decrement the node's reference count. In the fieldwatch case this isn't always true. The reason why is because the `value` node, and in some cases the "side-effect" node, is also used (and evaluated) by the load/store evaluator. Hence the load/store evaluators also decrement the reference counts for these children. In order to avoid double decrementing, we skip decrementing the reference count in the `rdbar` or `wrtbar` evaluator.

### Common FieldWatch Helper routine
After evaluating the required children to report the fieldwatch event, all the `rdbar` and `wrtbar` evaluators invoke the following routine in the common codegen:
```
rdWrtbarHelperForFieldWatch(TR::Node *node, TR::CodeGenerator *cg, TR::Register *sideEffectRegister, TR::Register *valueReg)
```

This routine gathers the data needed to report the fieldwatch, stores them in a snippet (more on the snippets later), then does the following things (in the specified order):
- If the field is unresolved, invoke the following platform specific routine that generates code (in an OOL section) to branch to a VM Helper routine that resolves the field:
```TR::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister)```
- Invoke the following platform specific routine to generate code (in an OOL section) to branch to a VM Helper that reports the fieldwatch event:
```TR::TreeEvaluator::readWriteBarrierHelperForFieldWatch(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister, TR::Register *valueReg)```

The details about the VM Helpers are available here: https://github.com/eclipse/openj9/issues/2668#issuecomment-414756016

### New data snippets for Fieldwatch
As mentioned above, we store the data needed to report a fieldwatch event in snippets. We have implemented two new snippet kinds that are robust and can be shared by all the code generators. These snippets are designed to hold data needed to report a fieldwatch event (i.e.  the `J9JITWatchedStaticFieldData` and `J9JITWatchedInstanceFieldData` structs needed by the VM Helper are overlayed on top of these snippets). 

There's one for static and one for instance fieldwatch events:
```J9WatchedStaticFieldSnippet```
```J9WatchedInstanceFieldSnippet```

These snippets inherit the `TR::Snippet` class and are implemented inside the common code generator. Platform specific routines don't have to worry about populating these snippets as it is now done at the common codegen level.

### AOT support
As mentioned in https://github.com/eclipse/openj9/issues/2668#issuecomment-414756016, AOT support currently does not exist. The new PR implements AOT support for fieldwatch. Most of the support is built into the Snippet kinds mentioned above, so not much work is required in the platform specific routines. Specifically, we add external relocations when emitting the snippet body for the following fields in the fieldwatch structures:
- `fieldAddress`
- `fieldClass`
- `j9Method`

In the platform specific routines, there are two scenarios where we must generate instructions that will be compatible for an AOT Compile:

1. The first scenario is when load the Constant Pool address into a register (the CP address needs to be passed in as an argument register when calling a VM Helper routine to resolve a field. This is done inside `generateFillInDataBlockSequenceForUnresolvedField`). 
Since the constant pool address is different for each JVM this address needs to be relocatable. As an example, on Z we use the following API to generate the correct instruction and add the correct external relocation record:
https://github.com/eclipse/omr/blob/5e6e968a9d930209eaa790f02e578535a1625bb7/compiler/z/codegen/S390GenerateInstructions.cpp#L2398-L2404

2. The platform specific routine `readWriteBarrierHelperForFieldWatch` needs to load in the `fieldClass` in order to check if fieldwatch is enabled (one of the bits in the class indicates if fieldwatch is enabled for the class). 
For resolved static fields this information is already available at compile time. In the non-AOT case we access the `fieldClass` directly at compile-time from the snippets described above. In the AOT case however, we must load the `fieldClass` into a register (ex. using a PC-Relative instruction). This way when the `fieldClass` is patched for a new JVM, we will load it correctly.


Having said all that, as things stand currently, AOT compilations are going to abort when fieldwatch is enabled. This is done because the VM Helpers for fieldwatch use VM Hooks. When these VM Hooks are enabled, an AOT compilation cannot run. We abort the compilation here:
https://github.com/eclipse/openj9/blob/3900caec909edcdd112d75a54e50ca4e4bd1ebe3/runtime/compiler/control/J9Options.cpp#L2140-L2163

The reason why we do this is because hook enablement is transitory. The hook state needs to be stored in the compiled body - something we don't do currently.

A separate item will need to investigate how to enable AOT Compiles when these VM Hooks are enabled. The support on the fieldwatch side should however be complete.


@cathyzhyi @fjeremic @gita-omr @AlenBadel @dsouzai FYI... Please let me know if you have any questions.

**Comment by gita-omr** [comment]
@dchopra001 thanks for the great description. I think it would be good to list the methods that need to be overriden by platforms. 

Also, do you mean that dwrtbar* and fwrtbar* evaluators are completely platform specific and don't use any of the common code that you wrote?

**Comment by dchopra001** [comment]
> I think it would be good to list the methods that need to be overriden by platforms.

Each platform requires its own implementation of the following methods:
```
TR::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister)
```

```
TR::TreeEvaluator::readWriteBarrierHelperForFieldWatch(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister, TR::Register *valueReg)
```
(Once the Pull Request is merged I'll come back and edit these comments to provide links in the codebase for any code I am referencing).

> Also, do you mean that dwrtbar* and fwrtbar* evaluators are completely platform specific and don't use any of the common code that you wrote?

Sorry I explained this incorrectly. The layout for these two evaluators is _similar_ to the ones in the common codegen. So they use the fieldwatch helper routine in the common codegen: `rdWrtbarHelperForFieldWatch`. I'll update the original comment.

**Comment by gita-omr** [comment]
Oh, my fault: you did list the platform specific methods already!

**Comment by dchopra001** [comment]
> Oh, my fault: you did list the platform specific methods already!

I missed one routine (`generateReportFieldAccessOutlinedInstructions`). I have updated the comment.

**Comment by gita-omr** [comment]
It's cosmetic, but I think it would be helpful if all three platform specific routines started with name 'generate.." This way it's easier to see which ones are common and which are platform specific.

**Comment by gita-omr** [comment]
Could we add a link to the design that describes various shapes of the rdbar and wrtbar trees? It's a bit hard to follow the code without knowing the shape. For example, static wrtbar seems to be different from others. Also, is sideEffect child always the last child?

**Comment by dchopra001** [comment]
The design of the IL is here: https://github.com/eclipse/omr/issues/2812#issue-345782313 and the corresponding pull request to introduce the OpCodes here: https://github.com/eclipse/omr/pull/2875

**Comment by gita-omr** [comment]
@AlenBadel : fyi (unless you already know it) - we cannot rely anymore on the fact that sideEffect child of *wrdbar (direct write) is an indirect load from J9Class and get it by doing node->getSecondChild()->getFristChild().

Other platforms now load J9Class from the sideEffectNode (second child). For example on z:

generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, fieldClassReg, generateS390MemoryReference(sideEffectRegister, cg->comp()->fej9()->getOffsetOfClassFromJavaLangClassField(), cg));

**Comment by AlenBadel** [comment]
In reaction to https://github.com/eclipse/openj9/issues/2668#issuecomment-493212820 

There are a few additional changes required to make this implementation sane for use on Power: 

1. Power does not have the ability of direct addressing, rather it uses relative addressing. Power uses a pseudo table of contents (pTOC), which is a buffer that stores addresses and constants for quick access. If the pTOC buffer overflows, then Power materialises the address in several instructions. The pTOC index, or the instructions generated are saved within the snippet class and referenced whenever access to the snippet is required.

To support this, placeholders within the snippets are required. Power would also need to overload the emit stage of the snippet in order to correctly populate the instructions, or insert the address of the snippet into the pTOC. This is done in addition to J9Watched(Instance/Staic)FieldSnippet. 

To accomplish this Power will need to create it's own snippet J9PPCWatched(Instance/Static)FieldSnippet which inherits the respective J9Watched(Instance/Static)FieldSnippet.
      There is some precedence to this in PPC.
https://github.com/eclipse/openj9/blob/e4cd7e133fae195923aee64f15b6eaeb67e2452c/runtime/compiler/p/codegen/CallSnippet.hpp#L157

https://github.com/eclipse/openj9/blob/e4cd7e133fae195923aee64f15b6eaeb67e2452c/runtime/compiler/p/codegen/PPCPrivateLinkage.cpp#L2115

2. Power AOT does not currently support contiguous relocation for TR_ClassAddress. Support for a new sequence kind is necessary to support this. 
See: https://github.com/eclipse/openj9/pull/4743#discussion_r285282691

**Comment by gacholio** [comment]
AOT support is less important, as this is only applicable to FSD mode, which I think rarely makes it into the cache.

**Comment by dchopra001** [comment]
Based on https://github.com/eclipse/openj9/issues/2668#issuecomment-494517069, it'd be good to add a codegen query that creates and returns a `J9Watched(Instance/Static)FieldSnippet`. On X and Z we will have something like this:

```
J9Watched(Instance/Static)FieldSnippet *getFieldWatchSnippet(..)
{
// create and return new J9Watched(Instance/Static)FieldSnippet
}
```

And on Power we would have:
```
J9Watched(Instance/Static)FieldSnippet *getFieldWatchSnippet(..)
{
// create and return new J9PPCWatched(Instance/Static)FieldSnippet
}
```

**Comment by gita-omr** [comment]
Yes, if we use dynamic polymorphism. Do you think it's worth trying implementing it via extensible classes?

**Comment by dchopra001** [comment]
> Yes, if we use dynamic polymorphism. Do you think it's worth trying implementing it via extensible classes?

IMO it's not worth it as I don't see much of a benefit for fieldwatch snippets. Dynamic polymorphism is probably the best solution.

**Comment by gita-omr** [comment]
Yeah, I agree, just wanted to confirm. Thanks!

**Comment by fjeremic** [comment]
> There are a few additional changes required to make this implementation sane for use on Power:
> 
>     1. Power does not have the ability of direct addressing, rather it uses relative addressing. Power uses a pseudo table of contents (pTOC), which is a buffer that stores addresses and constants for quick access. If the pTOC buffer overflows, then Power materialises the address in several instructions. The pTOC index, or the instructions generated are saved within the snippet class and referenced whenever access to the snippet is required.
> 
> 
> To support this, placeholders within the snippets are required. Power would also need to overload the emitSnippetBody of the snippet in order to correctly populate the instructions, or insert the address of the snippet into the pTOC. This is done in addition to J9Watched(Instance/Staic)FieldSnippet.
> 
> To accomplish this Power will need to create it's own snippet J9PPCWatched(Instance/Static)FieldSnippet which inherits the respective J9Watched(Instance/Static)FieldSnippet.
> There is some precedence to this in PPC.

Why does the snippet need to know anything about patching? Isn't this the job of the code generator to create the proper relocations [1] so that the relocation logic can patch whatever instructions you end up generating?

Can you give examples of what instructions you would actually emit for pTOC vs. materialization of the address? Where in the code do we do this? I'd like to further understand this. In my mind the way this should work if you don't support relative address materialization is:

1. Generate a pseudo instruction whose semantics are "LOAD ADDRESS X INTO REGISTER Y"
2. At Binary Encoding time decide whether you have enough pTOC space and if we do then store X into the pTOC at compile time and emit an instruction which loads X indirectly via the pTOC into register Y. If we don't have pTOC space then emit the instructions (I assume these are load immediate and shift instructions) to materialize the address inline.
3. If we generated materilaization inline add an internal relocation [1] to relocate the label symbol as the snippets have not yet been emitted

That's it. Now the snippet no longer has to care about any sort of patching as that's the job of the relocation logic, and we can all share the same snippet across 3 codegens without any platform specifics. Is this feasible? Am I misinterpreting the issue we have on Power?

[1] https://github.com/eclipse/omr/blob/master/compiler/codegen/Relocation.hpp

**Comment by gita-omr** [comment]
@fjeremic you are right, in general it probably would be a nicer solution. But it's not how currently similar situations are handled on Power. For example, we don't have a pseudo instruction like that. The best way to understand it is to take a look at our fconstEvaluator (in OMR). Essentially, the snippet controls everything. We might change this in the future but definitely not for this feature. Since the code that @dchopra001 implemented is very modular, it should not be too hard to change the Power part if we finally decide to use your approach.

**Comment by dchopra001** [comment]
> It's cosmetic, but I think it would be helpful if all three platform specific routines started with name 'generate.." This way it's easier to see which ones are common and which are platform specific.

I have renamed `readWriteBarrierHelperForFieldWatch` to `generateTestAndReportFieldWatchInstructions` to more accurately represent what the routine is supposed to do, as well as to keep the naming consistent.

**Comment by AlenBadel** [comment]
Due to the addressing constraints of Power, we need to make a small change to `generateTestAndReportFieldWatchInstructions` and `generateFillInDataBlockSequenceForUnresolvedField`. To avoid loading the address of the snippet twice, an additional optional parameter to these two methods has been added. This will allow Power to only load the address of the snippet once, either in the Unresolved Sequence or in the main Test and Report sequence. 

`
   /*
    * Generate instructions for static/instance field access report.
    * @param dataSnippetRegister: Optional, can be used to pass the address of the snippet inside the register.  
    */
`

`
static void generateTestAndReportFieldWatchInstructions(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister, TR::Register *valueReg, TR::Register *dataSnippetRegister);
`

`
   /*
    * Generates instructions to fill in the J9JITWatchedStaticFieldData.fieldAddress, J9JITWatchedStaticFieldData.fieldClass for static fields,
    * and J9JITWatchedInstanceFieldData.offset for instance fields at runtime. Used for fieldwatch support.
    * @param dataSnippetRegister: Optional, can be used to pass the address of the snippet inside the register.  
    */
`

`
static void generateFillInDataBlockSequenceForUnresolvedField (TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister, TR::Register *dataSnippetRegister);
`

**Comment by cathyzhyi** [comment]
Closing the issue since the work has been completed.
